File: c/tree/binary_tree/tree_node.c
#include <stdio.h>
#include <stdlib.h>

#include "tree_node.h"
#include "../../queue/queue_node.h"

struct TreeNode* createTreeNode(int val, struct TreeNode* left, struct TreeNode* right) {
    struct TreeNode* newNode = (struct TreeNode*) malloc(sizeof(struct TreeNode));
    newNode->val = val;
    newNode->left = left;
    newNode->right = right;
    return newNode;
}

void freeAll(struct TreeNode* root) {
    if (root == NULL) {
        return;
    }

    freeAll(root->left);
    freeAll(root->right);
    
    free(root);
}

void printInOrder(struct TreeNode* root) {
    if (root == NULL) {
        return;
    }

    printInOrder(root->left);

    printf("%d ", root->val);

    printInOrder(root->right);
}

void printPreOrder(struct TreeNode* root) {
    if (root == NULL) {
        return;
    }

    printf("%d ", root->val);

    printPreOrder(root->left);

    printPreOrder(root->right);
}

void printPostOrder(struct TreeNode* root) {
    if (root == NULL) {
        return;
    }

    printPostOrder(root->left);

    printPostOrder(root->right);

    printf("%d ", root->val);
}

void printLevelOrder(struct TreeNode* root) {
    if (root == NULL) {
        return;
    }

    Queue* q = createQueue();
    enqueue(q, root);

    while (!isQueueEmpty(q)) {
        struct TreeNode* current = dequeue(q);

        printf("%d ", current->val);

        if (current->left != NULL) {
            enqueue(q, current->left);
        }

        if (current->right != NULL) {
            enqueue(q, current->right);
        }
    }

    freeQueue(q);
}

File: java/libraries/tree/binary_tree/TreeNode.java
package libraries.tree.binary_tree;

import java.util.LinkedList;
import java.util.Queue;

public class TreeNode {
  public int val;
  public TreeNode left;
  public TreeNode right;

  public TreeNode(int val) {
    this.val = val;
  }

  public TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }

  public static void printInOrder(TreeNode root) {
    if (root == null) {
      return;
    }

    printInOrder(root.left);

    System.out.println(root.val + " ");

    printInOrder(root.right);
  }

  public static void printPreOrder(TreeNode root) {
    if (root == null) {
      return;
    }

    System.out.println(root.val + " ");

    printPreOrder(root.left);

    printPreOrder(root.right);
  }

  public static void printPostOrder(TreeNode root) {
    if (root == null) {
      return;
    }

    printPostOrder(root.left);

    printPostOrder(root.right);

    System.out.println(root.val + " ");
  }

  public static void printLevelOrder(TreeNode root) {
    if (root == null) {
      return;
    }

    Queue<TreeNode> q = new LinkedList<>();
    q.add(root);

    while (!q.isEmpty()) {
      TreeNode current = q.poll();
      System.out.print(current.val + " ");

      if (current.left != null) {
        q.add(current.left);
      }

      if (current.right != null) {
        q.add(current.right);
      }
    }
  }
}
